### PR TITLE
improve error messages for invalid indexing configurations

### PR DESCRIPTION
**Pull request by cmacdonald**
This addresses the experience reported in #348 by @maxhenze

### CODE REVIEW

**Comment by seanmacavaney** [comment]
Along this line, a few situations I've seen students encounter recently that could use better indexing error messages:
 - Short circuit error message if a docno is encountered that's longer than the configured maximum. (Currently no error thrown, only know once you get very poor performance when retrieving).
 - Clearer error message when user passes non-dict objects to IterDictIndexer (it's caught in Java and is rather cryptic)

**Comment by cmacdonald** [comment]
> Clearer error message when user passes non-dict objects to IterDictIndexer (it's caught in Java and is rather cryptic)

I have introduced a check for this on the first document. There is a related issue (tracking at https://github.com/terrier-org/terrier-core/issues/215) of not propagating the Java exception back into Python by FLATJsonDocument that we can fix upstream for a future release.

> Short circuit error message if a docno is encountered that's longer than the configured maximum. (Currently no error thrown, only know once you get very poor performance when retrieving).

Actually, this is caused by  `'metaindex.compressed.crop.long' : 'true'` which you have for IterDictIndexer (presumably to make text cropping automatic) - see https://github.com/terrier-org/pyterrier/blob/master/pyterrier/index.py#L803. We could pass the meta lengths to FlatJSONDocumentIterator and check the validity of the first document? This wont handle all possible error modes (e.g. first document is fine, but latter document is not), but would help, right?

**Comment by seanmacavaney** [comment]
Awesome, thanks!

> This wont handle all possible error modes (e.g. first document is fine, but latter document is not), but would help, right?

Yeah, this would probably cover most cases and would be a huge help.

But cases where things are numbered (e.g., `%p1` ... `%p10`) could consistently fail. The error message in the passage case would be that a duplicate docno was found, and the root cause could be tricky to get to the bottom of without this check.

The reasoning for only checking the first one is that checking every docno would get expensive? I imagine that it wouldn't add all that much overhead, especially as compared to the JSON generation, crossing between Python/Java, doing the actual indexing, etc.

**Comment by cmacdonald** [comment]
Ok, I have adjusted FlatJSONDocumentIterator to handle first doc. I note that FlatJSONDocumentIterator is only used by the non-FIFO setting, ie Windows, so these improvements would not work on Linux/macOS. The fifo impl seems to do more things natively in Java, and I cant see easily how to fix this. 

Aside: the fifo impl is more complicated and possibly unnecessary if the default is threads=1.

**Comment by seanmacavaney** [comment]
> and I cant see easily how to fix this

Should be able to just peek at the first doc [here](https://github.com/terrier-org/pyterrier/blob/master/pyterrier/index.py#L950), no? If you like, I could maybe take a look. 

> Aside: the fifo impl is more complicated and possibly unnecessary if the default is threads=1.

I suspect it's far more efficient though. From what I've seen benchmarking jnius, it's costly to move between Python and Java (which the nofifo one needs to do for every document, while the fifo one does not).

**Comment by cmacdonald** [comment]
> If you like, I could maybe take a look.

Sure, go for it.

File: pyterrier/index.py [hunk 1]
```diff
@@ -849,7 +883,7 @@ def index(self, it, fields=('text',), meta=None, meta_lengths=None, threads=None
             #     {'docno' : 'd1', 'toks' : {'a' : 1, 'aa' : 2}}
             # ]
             
-            iter_docs = DocListIterator(it)
+            iter_docs = DocListIterator(self._filter_iterable(it, fields))
```
Comment (seanmacavaney) [inline pyterrier/index.py:886]: I'm not sure why `self._filter_iterable` wasn't here before, as it is in the other invocation below.

**Comment by seanmacavaney** [comment]
@cmacdonald -- updated! Mind taking a look when you have a chance.

I also changed the warning to an error in the case of docno. I feel that if your docnos are being truncated, it is always a problem and it's best to force the user to deal with it before indexing will start. Other fields do not matter as much, so they remain just a warning.

**Comment by cmacdonald** [comment]
lgtm, after discussion

**Pull request merged**